static_website_tls/dispatch.ml

```diff
@@ -1,56 +1,106 @@
-        (aux (Re_str.(split_delim (regexp_string "/") path)))
-    in
+  let log c fmt = Printf.ksprintf (C.log c) fmt
+
+  let read_fs fs name =
+    FS.size fs name >>= function
+    | `Error (FS.Unknown_key _) ->
+      Lwt.fail (Failure ("read " ^ name))
+    | `Ok size ->
+      FS.read fs name 0 (Int64.to_int size) >>= function
+      | `Error (FS.Unknown_key _) -> Lwt.fail (Failure ("read " ^ name))
+      | `Ok bufs -> Lwt.return (Cstruct.copyv bufs)
 
-    (* dispatch non-file URLs *)
-    let rec dispatcher = function
-      | [] | [""] -> dispatcher ["index.html"]
-      | segments ->
-        let path = String.concat "/" segments in
-        Lwt.catch (fun () ->
-          read_fs path
-          >>= fun body ->
-          S.respond_string ~status:`OK ~body ()
-          ) (fun exn ->
-            S.respond_not_found ()
-          )
+  (* dispatch files *)
+  let rec dispatcher fs ?header uri = match Uri.path uri with
+    | "" | "/" -> dispatcher fs ?header (Uri.with_path uri "index.html")
+    | path ->
+      let mimetype = Magic_mime.lookup path in
+      let headers = Cohttp.Header.add_opt header "content-type" mimetype in
+      Lwt.catch
+        (fun () ->
+           read_fs fs path >>= fun body ->
+           S.respond_string ~status:`OK ~body ~headers ())
+        (fun exn ->
+           S.respond_not_found ())
+
+  (* Redirect to the same address, but in https. *)
+  let redirect uri =
+    let new_uri = Uri.with_scheme uri (Some "https") in
+    let new_uri = Uri.with_port new_uri (Some 4433) in
+    let headers =
+      Cohttp.Header.init_with "location" (Uri.to_string new_uri)
     in
+    S.respond ~headers ~status:`Moved_permanently ~body:`Empty ()
 
-    (* HTTP callback *)
-    let callback conn_id request body =
+  let serve c flow f =
+    let callback (_, cid) request body =
       let uri = Cohttp.Request.uri request in
-      dispatcher (split_path uri)
+      let cid = Cohttp.Connection.to_string cid in
+      log c "[%s] serving %s." cid (Uri.to_string uri);
+      f uri
     in
-    let conn_closed (_,conn_id) =
-      let cid = Cohttp.Connection.to_string conn_id in
-      C.log c (Printf.sprintf "conn %s closed" cid)
+    let conn_closed (_,cid) =
+      let cid = Cohttp.Connection.to_string cid in
+      log c "[%s] closing." cid
     in
-    let tcp = `TCP 443 in
-    let tls = `TLS (Tls.Config.server (), tcp) in
-    http tls (S.make ~conn_closed ~callback ())
+    let http = S.make ~conn_closed ~callback () in
+    S.listen http flow
+
+end
+
+(* HTTPS *)
+module HTTPS
+    (C : CONSOLE) (S : STACKV4)
+    (DATA : KV_RO) (KEYS: KV_RO)
+    (Clock : CLOCK) =
+struct
+
+  module TCP  = S.TCPV4
+  module TLS  = Tls_mirage.Make (TCP)
+  module X509 = Tls_mirage.X509 (KEYS) (Clock)
+
+  module Http  = Cohttp_mirage.Server(TCP)
+  module Https = Cohttp_mirage.Server(TLS)
+
+  module Dispatch_http  = Dispatch(C)(DATA)(Http)
+  module Dispatch_https = Dispatch(C)(DATA)(Https)
```

Comment:
@samoht Someone need to make devices for this pile of functors at some point. :p